prow/cmd/pipeline/controller.go

```diff
@@ -352,6 +378,61 @@ func (c *controller) pipelineID(pj prowjobv1.ProwJob) (string, string, error) {
 	return id, url, nil
 }
 
+func createProwJobIdentifier(pj *prowjobv1.ProwJob) string {
+	var additionalIdentifier string
+	if pj.Spec.Refs != nil {
+		var pulls []int
+		for _, pull := range pj.Spec.Refs.Pulls {
+			pulls = append(pulls, pull.Number)
+		}
+		sort.Ints(pulls)
+		additionalIdentifier = fmt.Sprintf("%s/%s@%s %v", pj.Spec.Refs.Org, pj.Spec.Refs.Repo, pj.Spec.Refs.BaseRef, pulls)
+	}
+	return fmt.Sprintf("%s %s", pj.Spec.Job, additionalIdentifier)
+}
+
+func getFilteredProwJobs(id string, pjsToFilter []*prowjobv1.ProwJob) []*prowjobv1.ProwJob {
+	pjs := []*prowjobv1.ProwJob{}
+	for _, p := range pjsToFilter {
+		if runningState(p.Status.State) && id == createProwJobIdentifier(p) {
+			pjs = append(pjs, p)
+		}
+	}
+	sort.Slice(pjs, func(i, j int) bool {
+		return pjs[i].Status.StartTime.Before(&pjs[j].Status.StartTime)
+	})
+	return pjs
+}
+
+func abortDuplicatedProwJobs(c reconciler, pj *prowjobv1.ProwJob) (*prowjobv1.ProwJob, error) {
+	if !runningState(pj.Status.State) || pj.Spec.Agent != prowjobv1.TektonAgent {
+		return pj, nil
+	}
+	id := createProwJobIdentifier(pj)
+	prowJobsToFilter, err := c.listProwJobs(pj.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	pjs := getFilteredProwJobs(id, prowJobsToFilter)
+	// do not abort the newest prowjob
+	for i := 0; i < len(pjs)-1; i++ {
+		newpj := pjs[i].DeepCopy()
+		now := c.now()
+		newpj.Status.State = prowjobv1.AbortedState
+		newpj.Status.Description = descAborted
+		newpj.Status.CompletionTime = &now
+		newpj, err = c.patchProwJob(pjs[i], newpj)
+		if err != nil {
+			logrus.WithError(err).Error("failed abort prowJob")
+			continue
+		}
+		if id == createProwJobIdentifier(pjs[i]) && pj.Status.StartTime == pjs[i].Status.StartTime {
+			pj = newpj.DeepCopy()
+		}
```

Comment:
This is a special rare case. If the prowjob that is currently passed to the function is on the list of prowjobs to be changed and was updated... it means we have an older definition and we need to update it as later we will interact with tekton upon the actual and up-to-date information from that pj (there is a big switch-case later on that utilizes up to date info). So if pj passed to a function is aborted we need to upgrade it definition.

Reply:
Hm... ok. That explains 2) -  I think it would be useful to add a comment here about it (basically what you wrote). 

Still, 1) applies, right? And I wonder whether we could match on something more explicit than `StartTime`, why don't we match on `pj.ObjectMeta.Name` or `pj.ObjectMeta.UID`?

Reply:
good question, I think names should be the same so I can change it to name. I do not have an explanation for why I thought id and start time are better ¯\_(ツ)_/¯

Reply:
changed to Name